lib/Server.js

```diff
@@ -490,7 +490,29 @@ class Server {
       options.devMiddleware = {};
     }
 
-    // No need to normalize `headers`
+    // normalize `headers`
+    if (typeof options.headers !== "undefined") {
+      const allHeaders = {};
+
+      if (Array.isArray(options.headers)) {
+        options.headers.forEach((header) => {
+          allHeaders[header.key] = header.value;
+        });
+        options.headers = allHeaders;
+      }
+
+      if (typeof options.headers === "function") {
+        const returnedHeaders = options.headers();
+
+        if (Array.isArray(returnedHeaders)) {
+          returnedHeaders.forEach((header) => {
+            allHeaders[header.key] = header.value;
+          });
+
+          options.headers = allHeaders;
+        }
+      }
+    }
```

Comment:
We can't normalize function here, because https://github.com/webpack/webpack-dev-server/blob/6b6ace1e576e69976b00eaf42c18b61b2faf1345/lib/Server.js#L1867, ideally we should convert object to array with key/value, but if it is function run it https://github.com/webpack/webpack-dev-server/blob/6b6ace1e576e69976b00eaf42c18b61b2faf1345/lib/Server.js#L1867 and if developer return object convert it to array too, then write headers

Reply:
>  ideally we should convert object to array with key/value, but if it is function run it

you mean provided `headers: { "x-foo": "yes"}` in config, normalize it to `headers: [{ key: 'x-foo', value: "yes"}]` ?

Reply:
Yes

Reply:
Any object to array, plus if developer return object from function convert too

Reply:
I will refactor 👍🏻